File: src/lib/gap-analysis.ts
import { config } from "./config";

export interface EmiInput {
  outstandingAmount: number;
  interestRatePct: number;
  tenureMonths: number;
}

export interface Debt extends EmiInput {
  id: string;
}

export function calculateEmi({
  outstandingAmount,
  interestRatePct,
  tenureMonths,
}: EmiInput): number {
  const monthlyRate = interestRatePct / 12 / 100;
  if (monthlyRate === 0) {
    return outstandingAmount / tenureMonths;
  }
  const growth = Math.pow(1 + monthlyRate, tenureMonths);
  return (outstandingAmount * monthlyRate * growth) / (growth - 1);
}

export function prioritizeDebts(debts: Debt[]): string[] {
  return [...debts]
    .sort((a, b) => b.interestRatePct - a.interestRatePct)
    .map((debt) => debt.id);
}

export type EmergencyFundStatus = "inadequate" | "building" | "adequate";

export interface EmergencyFundInput {
  monthlyExpenses: number;
  currentSavings: number;
}

export interface EmergencyFundResult {
  target: number;
  current: number;
  status: EmergencyFundStatus;
}

// No v1 source for the status thresholds (see decisions/log.md); halfway
// to target is the dividing line between "inadequate" and "building".
export function computeEmergencyFund({
  monthlyExpenses,
  currentSavings,
}: EmergencyFundInput): EmergencyFundResult {
  const target = monthlyExpenses * config.emergency_fund_multiplier;
  const status: EmergencyFundStatus =
    currentSavings >= target
      ? "adequate"
      : currentSavings >= target / 2
        ? "building"
        : "inadequate";

  return { target, current: currentSavings, status };
}

export interface TermCoverInput {
  annualIncome: number;
  existingTermCoverAmount: number;
}

export interface TermCoverResult {
  recommendedTermCover: number;
  gap: number;
}

export function computeTermCoverGap({
  annualIncome,
  existingTermCoverAmount,
}: TermCoverInput): TermCoverResult {
  const recommendedTermCover = annualIncome * config.income_replacement_multiplier;
  return {
    recommendedTermCover,
    gap: Math.max(0, recommendedTermCover - existingTermCoverAmount),
  };
}

export interface HealthCoverInput {
  personalHealthCoverAmount: number;
  employerHealthCoverAmount: number;
}

export interface HealthCoverResult {
  effectiveExistingCover: number;
  recommendedHealthCover: number;
  gap: number;
}

// No v1 source for how personal + employer cover combine (see
// decisions/log.md); treated as a straight sum.
export function computeHealthCoverGap({
  personalHealthCoverAmount,
  employerHealthCoverAmount,
}: HealthCoverInput): HealthCoverResult {
  const effectiveExistingCover = personalHealthCoverAmount + employerHealthCoverAmount;
  const recommendedHealthCover = config.health_cover_baseline;
  return {
    effectiveExistingCover,
    recommendedHealthCover,
    gap: Math.max(0, recommendedHealthCover - effectiveExistingCover),
  };
}

export interface KpiInput {
  emergencyFundCurrent: number;
  emergencyFundTarget: number;
  existingTermCoverAmount: number;
  recommendedTermCover: number;
  effectiveExistingCover: number;
  recommendedHealthCover: number;
  monthlyIncome: number;
  monthlyExpenses: number;
  debts: Debt[];
}

export interface KpiResult {
  emergencyFundCoveragePct: number;
  termCoverAdequacyPct: number;
  healthCoverAdequacyPct: number;
  savingsRatePct: number;
  debtToIncomePct: number;
}

export function computeKpis({
  emergencyFundCurrent,
  emergencyFundTarget,
  existingTermCoverAmount,
  recommendedTermCover,
  effectiveExistingCover,
  recommendedHealthCover,
  monthlyIncome,
  monthlyExpenses,
  debts,
}: KpiInput): KpiResult {
  const totalEmi = debts.reduce((sum, debt) => sum + calculateEmi(debt), 0);

  return {
    emergencyFundCoveragePct: (emergencyFundCurrent / emergencyFundTarget) * 100,
    termCoverAdequacyPct: (existingTermCoverAmount / recommendedTermCover) * 100,
    healthCoverAdequacyPct: (effectiveExistingCover / recommendedHealthCover) * 100,
    savingsRatePct: ((monthlyIncome - monthlyExpenses) / monthlyIncome) * 100,
    debtToIncomePct: (totalEmi / monthlyIncome) * 100,
  };
}

export interface GapAnalysisInput {
  monthlyIncome: number;
  monthlyExpenses: number;
  currentSavings: number;
  existingTermCoverAmount: number;
  personalHealthCoverAmount: number;
  employerHealthCoverAmount: number;
  debts: Debt[];
}

export interface GapAnalysisResult {
  emergencyFundTarget: number;
  emergencyFundCurrent: number;
  emergencyFundStatus: EmergencyFundStatus;
  debtPriorityOrder: string[];
  termCoverGap: number;
  healthCoverGap: number;
  emergencyFundCoveragePct: number;
  termCoverAdequacyPct: number;
  healthCoverAdequacyPct: number;
  savingsRatePct: number;
  debtToIncomePct: number;
}

// Composes 4.1-4.4 (emergency fund, debt priority, term/health gap) with
// the 4.5 KPI layer into the shape gap_analysis_results persists.
export function computeGapAnalysis(input: GapAnalysisInput): GapAnalysisResult {
  const emergencyFund = computeEmergencyFund({
    monthlyExpenses: input.monthlyExpenses,
    currentSavings: input.currentSavings,
  });
  const termCover = computeTermCoverGap({
    annualIncome: input.monthlyIncome * 12,
    existingTermCoverAmount: input.existingTermCoverAmount,
  });
  const healthCover = computeHealthCoverGap({
    personalHealthCoverAmount: input.personalHealthCoverAmount,
    employerHealthCoverAmount: input.employerHealthCoverAmount,
  });
  const debtPriorityOrder = prioritizeDebts(input.debts);
  const kpis = computeKpis({
    emergencyFundCurrent: emergencyFund.current,
    emergencyFundTarget: emergencyFund.target,
    existingTermCoverAmount: input.existingTermCoverAmount,
    recommendedTermCover: termCover.recommendedTermCover,
    effectiveExistingCover: healthCover.effectiveExistingCover,
    recommendedHealthCover: healthCover.recommendedHealthCover,
    monthlyIncome: input.monthlyIncome,
    monthlyExpenses: input.monthlyExpenses,
    debts: input.debts,
  });

  return {
    emergencyFundTarget: emergencyFund.target,
    emergencyFundCurrent: emergencyFund.current,
    emergencyFundStatus: emergencyFund.status,
    debtPriorityOrder,
    termCoverGap: termCover.gap,
    healthCoverGap: healthCover.gap,
    ...kpis,
  };
}

File: backend/lib/gap_analysis.py
from lib.finance_config import (
    EMERGENCY_FUND_MULTIPLIER, INCOME_REPLACEMENT_MULTIPLIER, HEALTH_COVER_BASELINE,
)


def calculate_emi(outstanding_amount: float, interest_rate_pct: float, tenure_months: int) -> float:
    monthly_rate = interest_rate_pct / 12 / 100
    if monthly_rate == 0:
        return outstanding_amount / tenure_months
    growth = (1 + monthly_rate) ** tenure_months
    return (outstanding_amount * monthly_rate * growth) / (growth - 1)


def prioritize_debts(debts: list[dict]) -> list[str]:
    return [d["id"] for d in sorted(debts, key=lambda d: d["interest_rate_pct"], reverse=True)]


# No v1 source for the status thresholds (see decisions/log.md); halfway
# to target is the dividing line between "inadequate" and "building".
def compute_emergency_fund(monthly_expenses: float, current_savings: float) -> dict:
    target = monthly_expenses * EMERGENCY_FUND_MULTIPLIER
    if current_savings >= target:
        status = "adequate"
    elif current_savings >= target / 2:
        status = "building"
    else:
        status = "inadequate"
    return {"target": target, "current": current_savings, "status": status}


def compute_term_cover_gap(annual_income: float, existing_term_cover_amount: float) -> dict:
    recommended_term_cover = annual_income * INCOME_REPLACEMENT_MULTIPLIER
    return {
        "recommended_term_cover": recommended_term_cover,
        "gap": max(0, recommended_term_cover - existing_term_cover_amount),
    }


# No v1 source for how personal + employer cover combine (see
# decisions/log.md); treated as a straight sum.
def compute_health_cover_gap(personal_health_cover_amount: float, employer_health_cover_amount: float) -> dict:
    effective_existing_cover = personal_health_cover_amount + employer_health_cover_amount
    recommended_health_cover = HEALTH_COVER_BASELINE
    return {
        "effective_existing_cover": effective_existing_cover,
        "recommended_health_cover": recommended_health_cover,
        "gap": max(0, recommended_health_cover - effective_existing_cover),
    }


def compute_kpis(
    emergency_fund_current: float, emergency_fund_target: float,
    existing_term_cover_amount: float, recommended_term_cover: float,
    effective_existing_cover: float, recommended_health_cover: float,
    monthly_income: float, monthly_expenses: float, debts: list[dict],
) -> dict:
    total_emi = sum(calculate_emi(d["outstanding_amount"], d["interest_rate_pct"], d["tenure_months"]) for d in debts)
    return {
        "emergency_fund_coverage_pct": (emergency_fund_current / emergency_fund_target) * 100,
        "term_cover_adequacy_pct": (existing_term_cover_amount / recommended_term_cover) * 100,
        "health_cover_adequacy_pct": (effective_existing_cover / recommended_health_cover) * 100,
        "savings_rate_pct": ((monthly_income - monthly_expenses) / monthly_income) * 100,
        "debt_to_income_pct": (total_emi / monthly_income) * 100,
    }


# Composes 4.1-4.4 (emergency fund, debt priority, term/health gap) with
# the 4.5 KPI layer into the shape gap_analysis_results persists.
def compute_gap_analysis(
    monthly_income: float, monthly_expenses: float, current_savings: float,
    existing_term_cover_amount: float, personal_health_cover_amount: float,
    employer_health_cover_amount: float, debts: list[dict],
) -> dict:
    emergency_fund = compute_emergency_fund(monthly_expenses, current_savings)
    term_cover = compute_term_cover_gap(monthly_income * 12, existing_term_cover_amount)
    health_cover = compute_health_cover_gap(personal_health_cover_amount, employer_health_cover_amount)
    debt_priority_order = prioritize_debts(debts)
    kpis = compute_kpis(
        emergency_fund_current=emergency_fund["current"],
        emergency_fund_target=emergency_fund["target"],
        existing_term_cover_amount=existing_term_cover_amount,
        recommended_term_cover=term_cover["recommended_term_cover"],
        effective_existing_cover=health_cover["effective_existing_cover"],
        recommended_health_cover=health_cover["recommended_health_cover"],
        monthly_income=monthly_income,
        monthly_expenses=monthly_expenses,
        debts=debts,
    )
    return {
        "emergency_fund_target": emergency_fund["target"],
        "emergency_fund_current": emergency_fund["current"],
        "emergency_fund_status": emergency_fund["status"],
        "debt_priority_order": debt_priority_order,
        "term_cover_gap": term_cover["gap"],
        "health_cover_gap": health_cover["gap"],
        **kpis,
    }
